### PR TITLE
Develop new non combinational CPU model

### PR DESCRIPTION
**Pull request by jardhu**
This pull request implements a modified CPU model that accounts for memory stalling. In particular a new `PipelinedNonCombinCPU` class is added containing the design of this CPU.

Likewise new functionality was implemented for the testing drivers to allow the tester driver to specify what type of memory and ports are required for the test. This is used primarily to perform the instruction tests and full application tests for the non combinational CPU.

All existing test cases should default to using the combinational memory and combinational ports and retain the previous testing behavior, at least from what I tested.

I specifically pushed my changes to the same branch I had before (deyahruhd:integrate-new-memory) since this resulted from merging the combinational stuff in early. This should not be a problem as far as I can tell except for some minor merging issues.

At this very moment though, the new CPU still does not pass all of the tests. I did not have time to fully think through the timing of each stage this week, so instead I focused more on extending the testing functionality to support customization of the new memory modules and the new CPU in advance.

### CODE REVIEW

File: Top.datastore.snapshot.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (powerjg) [inline Top.datastore.snapshot.json:1]: Probably shouldn't have been checked in :)

File: src/main/scala/components/hazardmemstall.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * The hazard detection unit, modified for memory stalling
+ *
+ * Input:  rs1, the first source register number
+ * Input:  rs2, the first source register number
+ * Input:  idex_memread, true if the instruction in the ID/EX register is going to read from memory
+ * Input:  idex_rd, the register number of the destination register for the instruction in the ID/EX register
+ * Input:  exmem_taken, if true, then we are using the nextpc in the EX/MEM register, *not* pc+4.
+ * Input:  imem_good, high if instruction memory is idle and ready.
```
Comment (powerjg) [inline src/main/scala/components/hazardmemstall.scala:15]: I thought "good" meant that the value on the output line was valid.

File: src/main/scala/components/memory/base-memory-components.scala [hunk 3]
```diff
@@ -8,40 +8,51 @@ import chisel3.util.experimental.loadMemoryFromFile
 /**
   * Base class for all modular backing memories. Simply declares the IO and the memory file.
   */
-class BaseDualPortedMemory(size: Int, memfile: String) extends Module {
+abstract class BaseDualPortedMemory(size: Int, memfile: String) extends Module {
   def wireMemory (imem: BaseIMemPort, dmem: BaseDMemPort): Unit = {
     // Connect memory imem IO to dmem accessor
-    this.io.imem.request <> imem.io.request
-    imem.io.response <> this.io.imem.response
+    this.io.imem.request <> imem.io.bus.request
```
Comment (powerjg) [inline src/main/scala/components/memory/base-memory-components.scala:14]: Maybe this should be in a separate patch? It's not clear how this relates to the CPU design.

File: src/main/scala/components/memory/memory-noncombin-ports.scala [hunk 4]
```diff
@@ -17,11 +17,22 @@ class OutstandingReq extends Bundle {
...
  * The instruction memory port. Since both the combinational and noncombinational instruction ports just issue
- * read requests in the same way both ports have the same implementation
+ * read requests in the same way both ports share the same implementation
  *
  * The I/O for this module is defined in [[IMemPortIO]].
  */
 class INonCombinMemPort extends ICombinMemPort {
+  // To fake combinational writes
+  val imemBusy  = RegInit (false.B)
+  imemBusy := DontCare
```
Comment (powerjg) [inline src/main/scala/components/memory/memory-noncombin-ports.scala:27]: This line worries me. Don't we want imemBusy to stay high or low until it changes? What does it mean to write a register with DontCare?

File: src/main/scala/components/memory/memory-noncombin-ports.scala [hunk 5]
```diff
@@ -17,11 +17,22 @@ class OutstandingReq extends Bundle {
...
+
+  when (io.pipeline.valid) {
+    imemBusy := true.B
+  } .elsewhen (io.bus.response.valid) {
+    imemBusy := false.B
+  }
+
+  io.pipeline.good := (~ imemBusy || io.bus.response.valid)
```
Comment (powerjg) [inline src/main/scala/components/memory/memory-noncombin-ports.scala:35]: I don't quite follow this logic. imemBusy will always be false when io.bus.response.valid is true, right? Also, it will be dontcare otherwise?

File: src/main/scala/components/memory/memory-noncombin-ports.scala [hunk 6]
```diff
@@ -30,6 +41,17 @@ class INonCombinMemPort extends ICombinMemPort {
  * The I/O for this module is defined in [[DMemPortIO]].
  */
 class DNonCombinMemPort extends BaseDMemPort {
+  // To fake combinational writes
+  val dmemBusy  = Reg (new Bool)
+  dmemBusy := DontCare
+
+  io.pipeline.good := (! dmemBusy || io.bus.response.valid)
```
Comment (powerjg) [inline src/main/scala/components/memory/memory-noncombin-ports.scala:48]: Same as above. Also, let's stick with `!` instead of `~` (see above). This is a *logical* statement. By using bitwise invert, if the size of dmemBusy changes a bug could be introduced.

File: src/main/scala/components/memory/memory-noncombin-ports.scala [hunk 7]
```diff
@@ -30,6 +41,17 @@ class INonCombinMemPort extends ICombinMemPort {
  * The I/O for this module is defined in [[DMemPortIO]].
  */
 class DNonCombinMemPort extends BaseDMemPort {
+  // To fake combinational writes
+  val dmemBusy  = Reg (new Bool)
```
Comment (powerjg) [inline src/main/scala/components/memory/memory-noncombin-ports.scala:45]: Why not RegInit like above?

File: src/main/scala/components/memory/base-memory-components.scala [hunk 8]
```diff
@@ -8,40 +8,51 @@ import chisel3.util.experimental.loadMemoryFromFile
...
 }
 
 /**
   * Base class for all data ports. Simply declares the IO.
   */
-class BaseDMemPort extends Module {
-  val io = IO (new DMemPortIO)
-  io      := DontCare
-  io.good := io.response.valid
+abstract class BaseDMemPort extends Module {
```
Comment (powerjg) [inline src/main/scala/components/memory/base-memory-components.scala:48]: I think adding the base ports should be in a separate PR / rebased commit.

File: src/main/scala/components/memory/memory.scala [hunk 9]
```diff
@@ -31,10 +31,15 @@ class DualPortedCombinMemory(size: Int, memfile: String) extends BaseDualPortedM
     // We should only be expecting a read from instruction memory
     assert(request.operation === Read)
     // Check that address is pointing to a valid location in memory
-    assert (request.address < size.U)
 
-    io.imem.response.valid        := true.B
-    io.imem.response.bits.data    := memory(request.address >> 2)
+    // TODO: Revert this back to the assert form "assert (request.address < size.U)"
```
Comment (powerjg) [inline src/main/scala/components/memory/memory.scala:35]: Can you create an issue with some more information about this TODO?

File: src/main/scala/components/memory/memory.scala [hunk 10]
```diff
@@ -53,147 +60,14 @@ class DualPortedCombinMemory(size: Int, memfile: String) extends BaseDualPortedM
...
-        data := readdata & ~(0xffff.U << (offset * 8.U))
-      }
-      memory(io.dmem.address >> 2) := data | (io.dmem.writedata << (offset * 8.U))
-    }.otherwise {
-      memory(io.dmem.address >> 2) := io.dmem.writedata
-    }
-  }
-}
-
+}
```
Comment (powerjg) [inline src/main/scala/components/memory/memory.scala:73]: You should set your editor to put newlines at the end of files :)

File: src/main/scala/pipelined/cpu-bp.scala [hunk 11]
```diff
@@ -387,6 +391,9 @@ class PipelinedCPUBP(implicit val conf: CPUConfig) extends Module {
   mem_wb.readdata   := io.dmem.readdata
   mem_wb.wbcontrol  := ex_mem.wbcontrol
 
+  // Stall pipeline if neither instruction nor data memory are ready
+  val memStall = ~(io.imem.good && io.dmem.good)
```
Comment (powerjg) [inline src/main/scala/pipelined/cpu-bp.scala:395]: Required for this design?

**Review by powerjg — COMMENTED** [review]
We need two documentation items:
1) What is the new class hierarchy?
2) What does the new pipeline look like?

Otherwise, I have mostly small comments below. I think it would be good to split this into to main changes. 1) Update the ports with pipeline/bus and update the CPU to be abstract (this could be 2 changes). 2) Implement new CPU

File: src/main/scala/components/memory/memory-port-io.scala [hunk 12]
```diff
@@ -7,32 +7,20 @@ import chisel3.util._
...
- *           an AsyncMemIO's response output, and should not be connected to anything else in any circumstance
- *           (or things will possibly break)
- *   Output: request, a DecoupledIO that delivers a request from a memory port to memory. This is primarily
- *           meant for connecting to an AsynMemIO's request input, and should not be connected to anything else
  */
 class MemPortIO extends Bundle {
   // Pipeline <=> Port
   val address  = Input(UInt(32.W))
   val valid    = Input(Bool())
   val good     = Output(Bool())
```
Comment (powerjg) [inline src/main/scala/components/memory/memory-port-io.scala:23]: Add a ready bit to replace what good used to do.

File: src/main/scala/components/hazardmemstall.scala (hunk 2 above)
Comment (jardhu) [reply src/main/scala/components/hazardmemstall.scala:15]: It had to be changed, I will push a commit with a ready output representing the same signal. This will necessarily be a change in the stable interfaces

File: src/main/scala/components/memory/memory-noncombin-ports.scala (hunk 7 above)
Comment (jardhu) [reply src/main/scala/components/memory/memory-noncombin-ports.scala:45]: It was a forgotten change from when I was trying to fix buggy behavior with these particular registers. It relates to another comment you had about the register being connected to DontCare.

File: src/main/scala/components/memory/memory-noncombin-ports.scala (hunk 4 above)
Comment (jardhu) [reply src/main/scala/components/memory/memory-noncombin-ports.scala:27]: This required a bit of testing but wiring a register with DontCare specifically makes it listen for any incoming signal on that cycle, and its necessary for the register to properly store values. Without the DontCare I believe as soon as it receives a high it will lock to being high. I'll elaborate with a timing chart similar to what I was reading while debugging it: 

Suppose the register starts with low and the write value alternates between low and high every cycle.
Without the DontCare line, the resulting register value will update like:
```
Cycle # | Register Val | Write Val
-------------------------------------
      1 |       0      |     0    
      2 |       0      |     1    
      3 |       1      |     0    
      4 |       1      |     1    
      5 |       1      |     0    
      6 |       1      |     1    
...
```
And if we initially write it with DontCare:
```
Cycle # | Register Val | Write Val
-------------------------------------
      1 |       0      |     0    
      2 |       0      |     1    
      3 |       1      |     0    
      4 |       0      |     1    
      5 |       1      |     0    
      6 |       0      |     1    
...
```
It appears to be a weird quirk with Chisel. All I know is that this line is required and removing it causes the register to act weird like in the first example.

File: src/main/scala/components/memory/memory-noncombin-ports.scala (hunk 5 above)
Comment (jardhu) [reply src/main/scala/components/memory/memory-noncombin-ports.scala:35]: When the non-combinational instruction port sends a request it will set imemBusy to high. A number of cycles later the memory bus will respond with true on the valid signal, thus implicitly indicating that it is ready for another request, but on that specific cycle ``imemBusy`` will also be high and needs one more cycle to be updated to false. 

So I OR'd the two signals together in this case. The memory will take one extra cycle to perform operations otherwise if we just pay attention to ``imemBusy``.

File: src/main/scala/components/memory/base-memory-components.scala (hunk 3 above)
Comment (jardhu) [reply src/main/scala/components/memory/base-memory-components.scala:14]: It shouldn't be appearing in the pull request, because these were changes that were already merged in from the previous PR. 

In hindsight I probably should have made a new branch to submit a pull request, but since this CPU design was an "extension" from the previous pull request. I thought it would be possible to push my new changes onto the same branch and have GitHub display just my new changes.

File: src/main/scala/pipelined/cpu-bp.scala (hunk 11 above)
Comment (jardhu) [reply src/main/scala/pipelined/cpu-bp.scala:395]: You're correct, I'll rebase this line out.

File: src/main/scala/components/memory/memory.scala (hunk 10 above)
Comment (jardhu) [reply src/main/scala/components/memory/memory.scala:73]: That's strange, it normally does.

**Comment by jardhu** [comment]
I am going to move the commits to this pull request into a new branch on my local repo. I believe since I did not merge master into my working branch beforehand, all of the commits that were already merged in before with the combinational memory are being counted as separate ones.

Obviously we do not want this as it will just pollute the commit history and possibly cause a lot of merging issues, so I'll just close this PR and create a new one using my new branch once I transfer everything over.
